### PR TITLE
Configuration API response to use dictionaries - based on PR #4695

### LINKED ISSUE #4556: Converting configuration API response from an array to dictionary.
## In what area(s)?

> /area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release
## Describe the proposal

Make all configuration APIs in all the SDKs to return a dictionary rather than an array.

In an array, we will have to iterate to get the value needed. But in the dictionary, we can get the value based on the key.

We are already passing the keys to get the configuration or subscribe to the configuration. From a dictionary, we can easily get the value of the config for whichever key we need. 

This will be an SDK level change. Once the dapr returns the list or array of objects, we can create a dictionary out of the response and return the dictionary to the user application.

For example:
This is an array/list of objects which is the response of the config API.

```
{
  [
      {
         key: "key1",
         value: "value1",
         version: "1.7",
         metadata: {}
      },
      {
         key: "key2",
         value: "value2",
         version: "1.7",
         metadata: {}
      },
  ] 
}
```

The proposal is to convert this list into a dictionary(key -> value):

```
"key1":  {
         key: "key1",
         value: "value1",
         version: "1.7",
         metadata: {}
      }
"key2":  {
         key: "key2",
         value: "value2",
         version: "1.7",
         metadata: {}
      }
```

In the above transformation, we can directly get the value when we know the key.

Docs and SDKs issues:

Js SDK:  * https://github.com/dapr/js-sdk/issues/283
Python SDK: * https://github.com/dapr/python-sdk/issues/411
Java SDK: * https://github.com/dapr/java-sdk/issues/746
Dotnet SDK: * https://github.com/dapr/dotnet-sdk/issues/884
Go SDK: * https://github.com/dapr/go-sdk/issues/291

Docs: * https://github.com/dapr/docs/issues/2484

### PR DESCRIPTION
**Pull request by berndverst**
# Description

This includes all commits from PR #4695 from @beiwei30 - thanks for the contributions! To speed up getting this feature merged I opened this new PR with all requires fixes / changes.

Fixes #4556
Fixes #4695 
Fixes #5071 

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/grpc/api.go [hunk 1]
```diff
@@ -1580,14 +1580,13 @@ func (a *api) GetConfigurationAlpha1(ctx context.Context, in *runtimev1pb.GetCon
 		return &runtimev1pb.GetConfigurationResponse{}, err
 	}
 
-	cachedItems := make([]*commonv1pb.ConfigurationItem, 0)
-	for _, v := range getResponse.Items {
-		cachedItems = append(cachedItems, &commonv1pb.ConfigurationItem{
-			Key:      v.Key,
+	cachedItems := make(map[string]*commonv1pb.ConfigurationItem, len(getResponse.Items))
```
Comment (mukundansundar) [inline pkg/grpc/api.go:1583]: is there a reason for this to be a pointer to ConfigItem?

**Review by mukundansundar — COMMENTED** [review]
LGTM. just one question ...

File: pkg/grpc/api.go (hunk 1 above)
Comment (berndverst) [reply pkg/grpc/api.go:1583]: @mukundansundar this is not my PR. This is from @beiwei30 - I don't know. I just dealt with the messy merge conflicts, missing tests and changes, and other broken things. If maintainers don't want it to be a pointer, please go ahead and fix up this PR here.

**Comment by berndverst** [comment]
@mukundansundar now this PR is failing because I rebased against latest dapr/dapr which introduces a change from @ItalyPaleAle..

we have about 5 conflicting changes with circular dependencies in flight. It's a complete mess.

File: pkg/grpc/api.go (hunk 1 above)
Comment (mukundansundar) [reply pkg/grpc/api.go:1583]: @dapr/maintainers-dapr any thoughts on avoiding pointers unless necessary?

**Comment by mukundansundar** [comment]
> @mukundansundar now this PR is failing because I rebased against latest dapr/dapr which introduces a change from @ItalyPaleAle..
> 
> we have about 5 conflicting changes with circular dependencies in flight. It's a complete mess.

wow ... all of them are failing! seems type mismatch ... probably use a prior commit from components-contrib to finish this PR and later a PR to use latest contrib ?

**Comment by berndverst** [comment]
@mukundansundar I think I know how to fix.. but it's a mess. Stay tuned.

**Review by artursouza — APPROVED** [review]

**Comment by berndverst** [comment]
Hopefully it's fixed now -- I had to get a change from @ItalyPaleAle into contrib and then import the latest contrib into this PR.

**Review by artursouza — APPROVED** [review]

**Pull request merged**